src/users.rs

```diff
@@ -0,0 +1,243 @@
+        &digest::SHA512,
+        N_ITER,
+        &salt,
+        password.as_bytes(),
+        &mut pbkdf2_hash,
+    );
+    pbkdf2_hash.to_vec()
+}
+
+// Verifies a password and salt against a hash
+// using the pbkdf2 algorithm
+fn verify_password(
+    password: String,
+    salt: Vec<u8>,
+    pbkdf2_hash: Vec<u8>,
+) -> Result<(), Unspecified> {
+    pbkdf2::verify(
+        &digest::SHA512,
+        N_ITER,
+        &salt,
+        password.as_bytes(),
+        &pbkdf2_hash,
+    )
+}
+
+// creates a sqlite DB if it does not exist and initializes the user table
+// returns the connection
+fn connection() -> Result<rusqlite::Connection, Box<dyn Error>> {
+    let conn = Connection::open("sqlite.db").expect("db conn fail");
+
+    // create the users table if it does not exist
+    match conn.execute(
+        "create table if not exists users (
+             id integer primary key,
+             name text not null unique,
+             salt text not null,
+             hash text not null,
+             active integer default 1 not null
+         );",
+        NO_PARAMS,
+    ) {
+        Ok(_) => {
+            return Ok(conn);
+        }
+        Err(_) => {
+            return Err(Box::new(UserError("Error Creating table".into())));
+        }
+    };
+}
+
+impl User {
+    // trait used to create a user
+    pub fn create(&mut self, password: String) -> Result<(), Box<dyn Error>> {
+        // Generates a salt for the user
+        let salt = get_salt();
+        self.salt = HEXUPPER.encode(&salt);
+        // Generates a hash from the salt and password
+        let hash = get_hash_from_password(password, salt);
+        self.hash = HEXUPPER.encode(&hash);
+        // Connect to the db
+        match connection() {
+            Ok(conn) => {
+                // Insert User into the db
+                match conn.execute(
+                    "INSERT INTO users (name, salt, hash, active) VALUES (?1, ?2, ?3, ?4)",
+                    params![self.name, self.salt, self.hash, self.active as i32],
+                ) {
+                    Ok(_) => {
+                        return Ok(());
+                    }
+                    Err(_) => {
+                        // There was an error inserting the user
+                        return Err(Box::new(UserError("Error Creating User".into())));
+                    }
+                }
+            }
+            _ => {
+                // Connection Error
+                return Err(Box::new(UserError("Connection Error".into())));
+            }
+        }
+    }
+    // Trait to Authorize the User
+    pub fn authorize(&mut self, password: String) -> Result<(), Box<dyn Error>> {
+        // Connect to DB
+        match connection() {
+            Ok(conn) => {
+                // Prepare Select statement
+                let mut stmt = conn
+                    .prepare("SELECT id, name, salt, hash, active FROM users WHERE name = :name LIMIT 1;")
+                    .expect("");
```

Comment:
You could leave this as it should never error, but I think `?` would be cleaner as other stuff can return an error anyway. Also you may as well call `unwrap` if you're not using the string.

Reply:
Same goes for the following `expects`. All of these will crash Trow if they ever occur.

Reply:
Will do this :)